File: runtimes/compose/src/main/java/com/styleconverter/runtime/core/renderer/BlockLabel.kt
package com.styleconverter.runtime.core.renderer

// The only Compose type the pure geometry below exposes: the shared ink
// colour the harness chrome paints its 1x1-px rects with.
import androidx.compose.ui.graphics.Color

/**
 * Block-font label GEOMETRY — the Compose copy of the cross-platform 5x7
 * glyph atlas math (BlockFont.gen.kt) that every platform's capture harness
 * rasterizes the component-name debug label with.
 *
 * WHAT THIS IS NOT (wave 51, PR A — the label-chrome move): the runtime no
 * longer draws the label. It used to be a composable placeholder INSIDE
 * the styled component, so the ink rode through the element's own paint
 * chain — offset by margin / relative insets / translate, rotated and
 * zoomed with it, clipped by filter layers, truncated at the CONTENT width
 * while web/iOS truncated at the border-box width — and the three captures
 * of one fixture disagreed on where (and whether) the label landed. The
 * label is now HARNESS CHROME: the capture app draws it as a SIBLING layer
 * over the whole capture root, after the component's paint, from the PLAIN
 * component name (`_` → space), at a fixed origin in the CAPTURE FRAME.
 * The normative rule lives in docs/DYNAMIC_CAPTURE.md, section "Harness
 * label chrome"; the Android drawer is `CaptureCanvas` in
 * apps/android-harness/…/screenshot/ScreenshotCaptureScreen.kt.
 *
 * Everything here is a PURE function (JVM-unit-testable without
 * Robolectric, the WptCaptureMode.kt pattern), and it is PUBLIC because the
 * harness is a different Gradle module (`app/build.gradle.kts`
 * `implementation(project(":runtime"))`) — a Kotlin `internal` object would
 * be invisible to it. BlockFont.gen.kt itself stays `internal`: nothing in
 * this object's signatures leaks it (`rects()` returns `List<Pair<Int,Int>>`).
 */
object BlockLabel {

    /** Cell-grid x origin IN THE CAPTURE FRAME: the first glyph cell starts
     *  8px right of the PNG's left edge — the shared cross-platform constant
     *  (web `left: 8px`, iOS `.offset(x: 8)`, Compose `drawRect` at x=8 on
     *  the capture root). Not relative to any component box any more — see
     *  docs/DYNAMIC_CAPTURE.md "Harness label chrome". */
    const val ORIGIN_X = 8

    /** Cell-grid y origin IN THE CAPTURE FRAME: glyph row 0 sits 6px below
     *  the PNG's top edge, so the 7 glyph rows (6..12) lie inside the 16px
     *  top padding band over the canvas ground and never overlap the border
     *  box of an in-flow root with non-negative margin-top / relative top
     *  (that box starts at y=16 on all three platforms). */
    const val ORIGIN_Y = 6

    /** Right-edge margin the truncation rule reserves against the FRAME
     *  width: characters are dropped until
     *  `ORIGIN_X + n*ADVANCE <= frameWidth - EDGE_MARGIN` (no ellipsis — the
     *  shared spec truncates silently). At the default 390px frame that is
     *  62 glyphs; at `CAPTURE_WIDTH=250` it is 39; below 22px it is 0. */
    const val EDGE_MARGIN = 8

    /**
     * The label ink: rgba(237, 237, 237, 179/255) — the debug-label colour
     * all three harnesses composite over the #1A1A2E ground. 0xED == 237 for
     * each RGB channel. Alpha ROUNDING CONTRACT (shared across platforms):
     * the historical 0.7 × 255 = 178.5 is rounded HALF-UP to 179 == 0xB3, so
     * the effective alpha every platform must rasterize with is exactly
     * 179/255 (~0.70196) → (174,174,180) over the ground; a platform that
     * used 178, or CSS `0.7` (which Chromium quantises one LSB darker to
     * (173,173,179)), diverges by one blend step on every lit pixel — which
     * is why the web harness moved to 179/255 in the same PR.
     */
    val COLOR = Color(0xB3EDEDED)

    /**
     * Shared label transform: UPPERCASE the input, and map any character
     * the atlas has no glyph for to '-' (the atlas only carries space, %,
     * -, ., 0-9, A-Z, _ — see BlockFont.GLYPHS). Uppercasing FIRST means
     * 'a'..'z' hit their capital glyphs instead of degrading to '-'.
     * The INPUT is the PLAIN component name with `_` replaced by a space —
     * the same string on web, iOS and Android (no text-transform, no
     * tab-size: those are the COMPONENT's styles, and the chrome is not
     * part of the component). Because this uppercases everything, a name
     * differs from its `text-transform: uppercase`d form by zero rects.
     */
    fun normalize(label: String): String = buildString(label.length) {
        // Per character: try the uppercased form against the atlas; every
        // miss (punctuation, accents, tabs, …) becomes the shared '-'.
        for (c in label) {
            val upper = c.uppercaseChar()
            append(if (BlockFont.GLYPHS.containsKey(upper)) upper else '-')
        }
    }

    /**
     * Truncation math (shared spec): keep the longest prefix of n chars
     * with `ORIGIN_X + n*ADVANCE <= frameWidthPx - EDGE_MARGIN`, i.e.
     * n = floor((frameWidth - 16) / 6), clamped to the label length.
     * [frameWidthPx] is the CAPTURE FRAME width — the number of pixels the
     * PNG is cut to (390 by default, `CAPTURE_WIDTH` otherwise) — never a
     * component or content-box width. Integer division IS the floor here
     * because the budget is coerced non-negative first (Kotlin `/`
     * truncates toward zero); widths 17..21 reach 0 through that division.
     */
    fun truncatedCount(labelLength: Int, frameWidthPx: Int): Int {
        // Horizontal pixel budget left for glyph cells once both the 8px
        // origin inset and the 8px right margin are reserved.
        val budget = frameWidthPx - ORIGIN_X - EDGE_MARGIN
        // No budget → no characters (never negative counts).
        if (budget <= 0) return 0
        // Whole advances that fit, never more than the label provides.
        return (budget / BlockFont.ADVANCE).coerceAtMost(labelLength)
    }

    /**
     * The full rect list for the first [count] characters of an already-
     * [normalize]d label: one (x, y) pair per SET bit, in CAPTURE-FRAME
     * pixels (so the first possible rect is at (8, 6) — the PNG's (8, 6)).
     * Cell origin for char i is x = ORIGIN_X + i*ADVANCE, y = ORIGIN_Y;
     * within a cell, bit 4 of the row int is the LEFTMOST column (col 0)
     * and row 0 is the top — the exact convention BlockFont.gen.kt
     * documents. This single pure function is BOTH what the harness paints
     * (one 1x1 `drawRect` per pair, integer offsets, no AA) and what the
     * unit suites pin, so tests and pixels cannot diverge.
     */
    fun rects(normalized: String, count: Int = normalized.length): List<Pair<Int, Int>> {
        // Pre-size roughly (≤ 35 bits/cell) — micro-optimization only.
        val out = ArrayList<Pair<Int, Int>>(count * 16)
        // Guard: callers pass a count from truncatedCount(), which is already
        // clamped, but never read past the string regardless (a stale or
        // over-large count must degrade to "whole label", not crash).
        for (i in 0 until count.coerceAtMost(normalized.length)) {
            // normalize() guarantees every char has a glyph → getValue is
            // total here (a miss would be a programming error, fail loud).
            val rows = BlockFont.GLYPHS.getValue(normalized[i])
            // Row 0 = top of the 7-row cell.
            for (row in 0 until BlockFont.CELL_H) {
                // The 5-bit row pattern for this scanline.
                val bitsRow = rows[row]
                // Col 0 = leftmost = bit (CELL_W - 1) = bit 4.
                for (col in 0 until BlockFont.CELL_W) {
                    // Test the bit for this column (shift the leftmost
                    // column down to bit 0).
                    if ((bitsRow shr (BlockFont.CELL_W - 1 - col)) and 1 == 1) {
                        // Emit the 1x1 rect's integer top-left.
                        out.add((ORIGIN_X + i * BlockFont.ADVANCE + col) to (ORIGIN_Y + row))
                    }
                }
            }
        }
        return out
    }

    /**
     * The CANONICAL atlas serialization — byte-identical to what
     * tools/visual/gen-block-font.mjs hashed to produce BlockFont.CHECKSUM:
     * glyphs sorted by codepoint as `<charCode>:<row,row,…>` joined with
     * ';', then `|<W>x<H>|adv<A>|lh<L>`. The pin test sha256s this string
     * (hex, first 16 chars) and compares against the embedded CHECKSUM —
     * the cross-platform drift guard that catches any hand-edit of the
     * generated constants or divergence between platform copies.
     */
    fun canonicalAtlasSerialization(): String {
        // Sort by codepoint — the generator's deterministic glyph order.
        val glyphPart = BlockFont.GLYPHS.entries
            .sortedBy { it.key.code }
            // `<charCode>:<rows joined with ','>` per glyph, ';'-joined.
            .joinToString(";") { (ch, rows) -> "${ch.code}:${rows.joinToString(",")}" }
        // Trailing geometry block: cell size, advance, line height.
        return glyphPart +
            "|${BlockFont.CELL_W}x${BlockFont.CELL_H}" +
            "|adv${BlockFont.ADVANCE}" +
            "|lh${BlockFont.LINE_HEIGHT}"
    }
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/Renderer/BlockLabel.swift
//
//  BlockLabel.swift
//  Renderer — the harness-label block-font canvas (applier campaign;
//  wave 51 PR (A): drawn as HARNESS CHROME, never by the runtime).
//
//  A deterministic 5x7 block-font raster: every glyph is a fixed bit grid
//  drawn as 1x1 integer-coordinate rect fills, so the label rasterizes
//  PIXEL-IDENTICALLY on all three platforms (real font stacks antialias
//  differently per platform and capped ~50 label-bearing fixtures at SSIM
//  0.90-0.949 however well the boxes matched). Real element text (the IR
//  `text` channel) is NOT routed here: it keeps PlaceholderLabel's full
//  typography path — fonts are the thing under test there.
//
//  WHO DRAWS IT (wave 51 PR (A); normative home: docs/DYNAMIC_CAPTURE.md,
//  section "Harness label chrome"): the capture HARNESS, as a sibling
//  overlay over the whole capture root — HarnessLabelChrome.swift, mounted
//  by apps/ios-harness/…/Screenshot/CaptureCanvas.swift right after the
//  root's `.background(…)`. Until (A) ComponentRenderer's leaf branch drew
//  this view INSIDE the styled element, so the ink rode the element's blend
//  group / opacity / filter / clip / transform and sat at the content-box
//  origin after the element's own layout — the platforms could not agree
//  on effect-bearing fixtures. The runtime now paints NO name label.
//
//  Geometry contract (shared verbatim with Compose's BlockLabel and web's
//  BlockFontLabel.ts — all three MUST produce byte-identical output):
//    * input = the PLAIN component name, underscores → spaces; transform
//      = uppercase, unknown glyph → '-'
//    * single line at integer origin (8, 6) in the CAPTURE FRAME; char i's
//      cell starts at x = 8 + i*ADVANCE. Rows 6..12 lie inside the 16px
//      top pad band, so on an in-flow root with non-negative margin-top
//      the ink never overlaps the border box (frame y >= 16).
//    * truncation against the FRAME width: largest n with
//      8 + n*ADVANCE <= frameWidth - 8 → 62 glyphs at 390, 39 at 250,
//      0 below 22 px (zero glyphs, logged once).
//    * color = rgba(237,237,237, 179/255); the CONTRACT is the COMPOSITED
//      byte, exactly (174,174,180) over #1A1A2E on all three (labelColor)
//    * NO antialiasing: integer coordinates only.
//
//  Glyph data comes from BlockFont.gen.swift (GENERATED — checksum
//  cb3c6e411c7b2859, pinned by BlockLabelTests against the canonical
//  serialization so the three embedded copies can never drift apart).
//

// SwiftUI for Canvas/Path/Color; CoreGraphics types (CGRect) ride along.
import SwiftUI

/// Pure layout math for the block label, split from the View so XCTest
/// pins transform/truncation/bit-geometry without a render surface (the
/// same pattern as StyleBuilder.minFloor / suppressesNamePlaceholder).
// public: the label contract's harness-visible geometry (design C23).
public enum BlockLabelLayout {

    /// Fixed label origin in the capture frame: 8px from the left edge.
    /// Part of the shared spec — web/Compose place the run at the same x.
    public static let originX: CGFloat = 8
    /// Fixed label origin in the capture frame: 6px from the top edge —
    /// inside the canvas's 16px top pad band, clear of the border box.
    public static let originY: CGFloat = 6
    /// Symmetric right margin: the truncation bound keeps the glyph run
    /// out of the last 8px of the frame (mirror of originX).
    public static let rightMargin: CGFloat = 8

    /// The label transform: uppercase the WHOLE string first (Swift's
    /// ICU-backed `uppercased()` matches JS `toUpperCase()` / Kotlin
    /// `uppercase()` for the ASCII names the harness emits), then map every
    /// unicode SCALAR the atlas has no glyph for to '-' — a visible stand-in,
    /// never a silent drop. Scalars, NOT `Character`s: web/tooling walk code
    /// points (`Array.from`) and Kotlin walks chars, so e + U+0301 must give
    /// TWO glyphs ('E','-') here too, not one grapheme '-' (BlockLabelTests).
    public static func glyphString(_ label: String) -> String {
        // Per-scalar atlas lookup AFTER the whole-string uppercase; every
        // atlas key is one ASCII scalar, so Character(s) is the exact key,
        // and '-' is guaranteed present in the atlas (BlockFont.gen.swift).
        return String(label.uppercased().unicodeScalars.map { s in
            BlockFont.glyphs[Character(s)] != nil ? Character(s) : "-"
        })
    }

    /// How many leading characters fit: largest n with
    /// 8 + n*ADVANCE <= width - 8 (no ellipsis — trailing chars just
    /// drop, per the shared spec). `componentWidth` is the FRAME width
    /// under PR (A) (the parameter keeps its historical name so the
    /// cross-platform math reads the same); nil = no bound, so the full
    /// run always fits (unit tests, fit-content callers).
    public static func truncatedCount(_ glyphCount: Int, componentWidth: CGFloat?) -> Int {
        // No statically-known width → no truncation by construction.
        guard let w = componentWidth else { return glyphCount }
        // n <= (W - leftInset - rightMargin) / ADVANCE, floored to an
        // integer character count and clamped to [0, glyphCount].
        let fit = Int(((w - originX - rightMargin) / CGFloat(BlockFont.advance)).rounded(.down))
        return max(0, min(glyphCount, fit))
    }

    /// One 1x1 rect per SET bit, in CANVAS-LOCAL coordinates (the canvas
    /// itself is offset to (originX, originY) by the view below, which
    /// puts char i's cell at frame-space x = 8 + i*ADVANCE, y = 6 exactly
    /// as the shared spec requires). Row 0 is the top row; bit 4 of each
    /// row int is the LEFTMOST of the 5 columns (generator contract,
    /// BlockFont.gen.swift header).
    public static func bitRects(_ glyphs: String) -> [CGRect] {
        // Flat rect list — order is irrelevant (all fills are the same
        // color), only the covered pixel set matters.
        var rects: [CGRect] = []
        for (i, ch) in glyphs.enumerated() {
            // Post-transform every char has a glyph; the guard is an
            // honest no-silent-fallthrough for direct (test) callers
            // that skip glyphString — unknown chars draw nothing.
            guard let rows = BlockFont.glyphs[ch] else { continue }
            // Cell origin advances one fixed-width cell per character.
            let cellX = CGFloat(i * BlockFont.advance)
            for (row, bits) in rows.enumerated() {
                // Column c reads bit (cellW-1-c): bit 4 = leftmost.
                for c in 0..<BlockFont.cellW where (bits >> (BlockFont.cellW - 1 - c)) & 1 == 1 {
                    // Integer 1x1 rect = exactly one full pixel at 1x —
                    // full coverage, so no antialiased edge can appear.
                    rects.append(CGRect(x: cellX + CGFloat(c), y: CGFloat(row),
                                        width: 1, height: 1))
                }
            }
        }
        return rects
    }
}

/// The block-font label view: a Canvas of 1x1 rect fills that collapses
/// to a 0×0 layout slot and hangs its ink at (8, 6) from that slot's
/// top-leading corner. Hosted ONLY by HarnessLabelChrome inside the
/// harness's `.overlay(alignment: .topLeading)` over the capture root, so
/// the slot's corner IS the frame origin and the ink lands at frame (8, 6).
// public: a public struct's synthesized memberwise init is INTERNAL, so
// the init below is spelled out (design C23; the ComponentHost precedent).
public struct BlockLabel: View {

    /// The PLAIN component name with underscores already replaced by
    /// spaces (HarnessLabelChrome.labelText).
    public let label: String

    /// The width the run truncates against — the capture FRAME width
    /// (CaptureCanvas.width) under PR (A). nil = no bound (unit tests).
    public let componentWidth: CGFloat?

    // public: explicit memberwise init for cross-module callers (the
    // harness app's plain `import StyleConverterRuntime`).
    public init(label: String, componentWidth: CGFloat?) {
        self.label = label
        self.componentWidth = componentWidth
    }

    /// rgba(237,237,237, 179/255). The CONTRACT is the COMPOSITED byte: ink
    /// over the #1A1A2E stage reads exactly (174,174,180) on all three
    /// platforms (design-record §2, docs/DYNAMIC_CAPTURE.md §5). 179/255 is
    /// spelled as the byte for cross-platform readability (Compose packs
    /// 0xB3EDEDED, 0xB3 = 179); measured on the Catalyst raster, 0.7
    /// composites to the same byte (skeptic-ios.md M-α), so it is not a
    /// rounding fix. Web spells it 0.706 (byte 180): Chromium rounds CSS
    /// alpha to 8 bits first and byte 179 lands one LSB dark (173,173,179).
    static let labelColor = Color(red: 237.0 / 255.0,
                                  green: 237.0 / 255.0,
                                  blue: 237.0 / 255.0,
                                  opacity: 179.0 / 255.0)

    /// The glyph run that actually draws: transform (uppercase, unknown →
    /// '-') then truncation against `width` (shared-spec formula). Also
    /// the ONE place the zero-glyph case is reported — a frame narrower
    /// than 22 px drops every glyph (design C5: zero glyphs, no crash; the
    /// Android twin logs the same case). Logged once per label so a blank
    /// band is explainable; a statement, so it lives outside the ViewBuilder.
    static func shownGlyphs(label: String, width: CGFloat?) -> String {
        let glyphs = BlockLabelLayout.glyphString(label)
        let count = BlockLabelLayout.truncatedCount(glyphs.count, componentWidth: width)
        if count == 0 && !glyphs.isEmpty {
            PropertyTracker.logOnce(
                key: "blocklabel.zero-glyphs.\(label)",
                message: "[BlockLabel] frame \(width.map { "\($0)" } ?? "nil")px "
                    + "too narrow for any glyph of \"\(label)\" — label dropped")
        }
        return String(glyphs.prefix(count))
    }

    // public: View protocol witness on a public type must be public.
    public var body: some View {
        // Transform + truncate once per render (pure apart from the
        // once-only zero-glyph log above).
        let shown = Self.shownGlyphs(label: label, width: componentWidth)
        // Precompute the pixel set outside the draw closure — the draw
        // closure then only replays fills (cheap + deterministic).
        let rects = BlockLabelLayout.bitRects(shown)
        // rendersAsynchronously: false — the harness captures via
        // ImageRenderer immediately after layout; async rasterization
        // could hand the capture an empty layer (wave-3 Canvas lesson).
        Canvas(rendersAsynchronously: false) { context, _ in
            // One fill per set bit, exactly as the shared spec words it.
            // Integer-aligned unit rects have full pixel coverage, so
            // the default antialiasing state can never soften an edge.
            for r in rects {
                context.fill(Path(r), with: .color(Self.labelColor))
            }
        }
        // ZERO LAYOUT FOOTPRINT: the label is chrome, not content — it
        // must never contribute to the capture's size. The canvas spans
        // the truncated run × LINE_HEIGHT …
        .frame(width: CGFloat(shown.count * BlockFont.advance),
               height: CGFloat(BlockFont.lineHeight))
        // … then the layout slot collapses to 0×0 and the (already-framed)
        // canvas hangs off its top-leading corner at the spec origin.
        // `overlay` content never influences the host's size (the wave-3
        // CaptureCanvas lesson), and Canvas draws are not clipped by the
        // zero frame — so under the harness overlay this is frame (8, 6).
        .offset(x: BlockLabelLayout.originX, y: BlockLabelLayout.originY)
        .frame(width: 0, height: 0, alignment: .topLeading)
    }
}
